### ISSUE TITLE
Sharp increase of compilation times for Android

### ISSUE THREAD

**Issue by panos-lunarg** [LunarG/gfxreconstruct#1044]
On current `dev`, a full build (`./gradlew build`) on my system takes a little over 8 minutes.
Picking a random commit from around June 2022 `./gradlew build` takes around 3 minutes and 20 seconds.
This is around x2 increase and I think it is worth the time looking into

**Comment by bradgrantham-lunarg** [comment]
Any ideas what the increase is?  We build for a bunch of platform variants, and I wonder if it's just a multiplication of variants.

**Comment by MarkY-LunarG** [comment]
I did 3 builds of the commit with Locke's change to update the header [commit d6dcd2baea4f98d4572cda3217197c4aa2505372].

Each one I remove the old build folders, and perform a gradle -c and then a gradle build:

```
    rm -rf android/framework/application/build
    rm -rf android/framework/decode/build
    rm -rf android/framework/encode/build
    rm -rf android/framework/format/build
    rm -rf android/framework/graphics/build
    rm -rf android/framework/util/build
    rm -rf android/layer/.cxx
    rm -rf android/layer/build
    rm -rf android/tools/build
    rm -rf android/tools/replay/build
    $GRADLE_HOME/gradle clean
    $GRADLE_HOME/gradle ${BUILD_TARGET}
```

I get the following times:
 - 148 seconds
 - 145 seconds
 - 142 seconds

I am running a laptop with 32gb of RAM, maybe you have less RAM and you could be paging now?

**Comment by panos-lunarg** [comment]
With a quick investigation it seems that release (`./gradlew assembleRelease`) build takes the longest with ~6:30 minutes wile debug build (`./gradlew assembleDebug`) takes 1-2 minutes.
Peeking in `htop` it seems that compilation of `framework/generated/generated_vulkan_replay_consumer.cpp` takes the longest time in release builds as it the one I see to linger when other files have been compiled.

Trying a random commit from June (96222921) yields 3 min for release build and 1:30 minutes for debug build.

I wanted to see the compilation times on Jenkins from other build jobs for comparison but I just realized that I cannot establish a vpn connection.

@MarkY-LunarG That's interesting. My laptop has 16gb of ram but it does seem to be enough as I can't see it spilling over to swap while compiling. I also tried cleaning everything up before compiling but this didn't seem to have any effect

**Comment by MarkY-LunarG** [comment]
I see a big increase with the recent changes by @locke-lunarg  on March 17.  For the debug build, we went from 116 seconds  (commit `6b12a2`) to 125+ seconds per build at (commit `e54abaf`), then jumped up to 145+ seconds at commit `ba97496`.

**Comment by andrew-lunarg** [comment]
> then jumped up to 145+ seconds at commit `ba97496`

We could make all those `unordered_map` instantiations all using one type: <handle, void*> and then turn all the template access functions in the header into simple wrappers that just throw away or cast their types and call into a shared non-inline type erased function whose body lives in the `.cpp` file. Just get a whole bunch of template code out of the header.

**Comment by wpaul-samsung** [comment]
I looked at the build times for a bit as well.  We're building for Android using our own GitHub workflows and are seeing ~15m on Linux and ~21m on Windows.  That's for assembleRelease.  I see in https://github.com/LunarG/gfxreconstruct/actions/runs/5415034542/jobs/9842918961 it took about 48m to build debug + release.

Obviously these times are much higher than seen locally by Mark and Panos.  Maybe the VMs used by workflows are just that much slower than physical machines.

Personally on my Windows laptop I am able to build Android release in ~16m.

I did see a ton of warnings from the builds which I have fixed locally by updating the gradle version to 7.5.1 and the Android gradle plugin version to 7.4.2.  Those new version mentions some other deprecations which I have also fixed locally.  The new versions did not seem to improve the build times though.  Let me know if you're interested and I can push my changes to a branch.

**Comment by panos-lunarg** [comment]
I did a lot of Android builds recently and apart from a couple of bottlenecks we have already detected (#1255 and #1268) I believe that I have found another one.
On my system the debug build of the file `generated_vulkan_replay_consumer.cpp` for `arm64-v8a` takes around 10 seconds but for release it takes around 5 minutes.
Responsible for the 95% of that is the line https://github.com/LunarG/gfxreconstruct/blob/dev/framework/generated/generated_vulkan_replay_consumer.cpp#L9211
nop'ing that line (removing the call to `GetPointer()`) drop the compilation down to seconds

**Comment by andrew-lunarg** [comment]
> nop'ing that line (removing the call to `GetPointer()`) drops the compilation down to seconds

The line does no work. Making it null or whatever you did must let the compiler eliminate the body of the following loop.

**Comment by andrew-lunarg** [comment]
How about replacing the hundreds of template instantiations of that huge function with one concrete huge function and hundreds of little template function instantiations which boil away? This will save us a lot of code space too.

Something like this untested:
```cpp

void InitializeOutputStructPNext(const size_t len, const VkBaseInStructure* input, VkBaseOutStructure* output);

// Template just for type coercion that boils away to no overhead:
template <typename T>
void InitializeOutputStructPNext(StructPointerDecoder<T> *decoder)
{
    if(decoder->IsNull()) return;
    size_t len = decoder->GetOutputLength();
    auto input = decoder->GetPointer();
    auto output = decoder->GetOutputPointer();

    InitializeOutputStructPNext(len, reinterpret_cast<const VkBaseInStructure*>(input), reinterpret_cast<VkBaseOutStructure*>(output));
}

// Concrete function too large to be likely selected for inlining into template above:
void InitializeOutputStructPNext(const size_t len, const VkBaseInStructure* input, VkBaseOutStructure* output)
{
    for( size_t i = 0 ; i < len; ++i )
    {
        const auto* in_pnext = reinterpret_cast<const VkBaseInStructure*>(input[i].pNext);
        if( in_pnext == nullptr ) continue;
        auto* output_struct = reinterpret_cast<VkBaseOutStructure*>(&output[i]);

        while(in_pnext)
        {
            switch(in_pnext->sType)
            {
                case VK_STRUCTURE_TYPE_BUFFER_MEMORY_BARRIER:
                {
                    output_struct->pNext = reinterpret_cast<VkBaseOutStructure*>(DecodeAllocator::Allocate<VkBufferMemoryBarrier>());
                    break;
                }
                case VK_STRUCTURE_TYPE_IMAGE_MEMORY_BARRIER:
                {
                    output_struct->pNext = reinterpret_cast<VkBaseOutStructure*>(DecodeAllocator::Allocate<VkImageMemoryBarrier>());
                    break;
                }

                /// ... same as current function for hundereds more lines
                
                case VK_STRUCTURE_TYPE_PHYSICAL_DEVICE_MESH_SHADER_PROPERTIES_EXT:
                {
                    output_struct->pNext = reinterpret_cast<VkBaseOutStructure*>(DecodeAllocator::Allocate<VkPhysicalDeviceMeshShaderPropertiesEXT>());
                    break;
                }
                default:
                    break;
            }
            output_struct = output_struct->pNext;
            output_struct->sType = in_pnext->sType;
            in_pnext = in_pnext->pNext;
        }
    }
}

**Comment by panos-lunarg** [comment]
```
void InitializeOutputStructPNext(const size_t len, const VkBaseInStructure* input, VkBaseOutStructure* output);
```
As far as I can tell `InitializeOutputStructPNext` can not be non tamplate, at least not in the way I understand your example, as you can not pass `VkBaseInStructure` and `VkBaseOutStructure` because `input[i]` and `output[i]` will be wrong.
I uploaded a draft PR (#1297) with simple pointer arithmetic in order to drop the template. Let me know if you come up with some smarter c++

**Comment by panos-lunarg** [comment]
This has been resolved by #1318. Closing
